File: 17.letter-combinations-of-a-phone-number.java
/*
 * @lc app=leetcode id=17 lang=java
 *
 * [17] Letter Combinations of a Phone Number
 */

// @lc code=start

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

class Solution {
    private List<String> results;
    private StringBuilder letters;
    private Map<Character, List<Character>> map = Map.of(
            '2', List.of('a', 'b', 'c'),
            '3', List.of('d', 'e', 'f'),
            '4', List.of('g', 'h', 'i'),
            '5', List.of('j', 'k', 'l'),
            '6', List.of('m', 'n', 'o'),
            '7', List.of('p', 'q', 'r', 's'),
            '8', List.of('t', 'u', 'v'),
            '9', List.of('w', 'x', 'y', 'z'));

    public List<String> letterCombinations(String digits) {
        if (digits.equals("")) {
            return new ArrayList<>();
        }
        results = new ArrayList<>();
        letters = new StringBuilder(digits.length());
        backtracking(digits.toCharArray(), 0);
        return results;
    }

    private void backtracking(char[] digits, int startIndex) {
        if (letters.length() == digits.length) {
            results.add(letters.toString());
            return;
        }
        List<Character> chars = map.get(digits[startIndex]);
        for (char ch : chars) {
            letters.append(ch);
            backtracking(digits, startIndex + 1);
            letters.deleteCharAt(letters.length() - 1);
        }
    }
}
// @lc code=end

File: 17.letter-combinations-of-a-phone-number.cpp
/*
 * @lc app=leetcode id=17 lang=cpp
 *
 * [17] Letter Combinations of a Phone Number
 */

// @lc code=start

#include <vector>
#include <unordered_map>
#include <string>

using namespace std;

class Solution {
private:
    unordered_map<char, vector<char>> keymappings = {
        {'2', {'a', 'b', 'c'}},
        {'3', {'d', 'e', 'f'}},
        {'4', {'g', 'h', 'i'}},
        {'5', {'j', 'k', 'l'}},
        {'6', {'m', 'n', 'o'}},
        {'7', {'p', 'q', 'r', 's'}},
        {'8', {'t', 'u', 'v'}},
        {'9', {'w', 'x', 'y', 'z'}},
    };

    string digits_;
    vector<string> results_;
    vector<char> bag_;

    void backtracking(string::iterator begin) {
        if (begin == digits_.end()) {
            if (bag_.size() == digits_.size()) {
                results_.push_back(string {bag_.begin(), bag_.end()});
            }
            return;
        }
        for (auto iter = begin; iter != digits_.end(); ++iter) {
            for (auto ch : keymappings[*iter]) {
                bag_.push_back(ch);
                backtracking(iter + 1);
                bag_.pop_back();
            }
        }
    }

public:
    vector<string> letterCombinations(string digits) {
        if (digits == "") return {};
        digits_ = digits;
        backtracking(digits_.begin());
        return results_;
    }

};
// @lc code=end
